### PR TITLE
Spark 3.4: Add write options to override the compression properties of the table

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?
Flink has supported to use write option to override the compression properties of the table.
I refer to the pull request #6049 and then I add write options to override the compression properties of the table for Spark 3.4.

### Why are the changes needed?
First, there exist some old tables using `gzip` compression codec in the  production environment.  If we use `zstd` compression codec to rewrite the data, we can reduce the data volume and reduce the cost of storage.
Second, this pr is also meaning after we choose `zstd` as default compression codec.  Because we can choose different compression levels when we write the Iceberg data and rewrite the Iceberg data if we have this pr. 

### Does this PR introduce _any_ user-facing change?
Yes. 
This pr introduces the config option ` compression-codec`, `compression-level` and `compression-strategy`. (document added)
This pr introduces the Spark config properties `spark.sql.iceberg.write.compression-codec`,  `spark.sql.iceberg.write.compression-level` and `spark.sql.iceberg.write.compression-strategy`. (I will add the document after this pr is merged.)

### How was this patch tested?
Add a new ut and manual verification.
I use the ut to verify that the compression codec is correct.
I verify the compression config option is correct by hand.
企业微信截图_93ca4fed-cd04-4315-819b-a56afdea106c

企业微信截图_fdb11295-8800-4ccc-ab5a-ec0abf12e913

### CODE REVIEW

**Comment by jerqi** [comment]
@chenjunjiedada @ConeyLiu Could you help me review this pr?

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    tableProperties.put(PARQUET_COMPRESSION, "gzip");
+    tableProperties.put(AVRO_COMPRESSION, "gzip");
+    tableProperties.put(ORC_COMPRESSION, "zlib");
+    Table table =
+        tables.create(SCHEMA, PartitionSpec.unpartitioned(), tableProperties, location.toString());
+    List<SimpleRecord> expectedOrigin = Lists.newArrayList();
+    for (int i = 0; i < 1000; i++) {
+      expectedOrigin.add(new SimpleRecord(1, "hello world" + i));
+    }
+    Dataset<Row> df = spark.createDataFrame(expectedOrigin, SimpleRecord.class);
```
Comment (chenjunjiedada) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:138]: nit： blank line after block.

**Comment by chenjunjiedada** [comment]
@jerqi Could you add some doc like in #6049 since `compression-level` and `compression-strategy` are targeted for different formats separately?  Also, it would be better to add a description of the default value and the available options. Others LGTM. Thanks for the contribution.

**Comment by jerqi** [comment]
> @jerqi Could you add some doc like in #6049 since `compression-level` and `compression-strategy` are targeted for different formats separately? Also, it would be better to add a description of the default value and the available options. Others LGTM. Thanks for the contribution.

Thanks for your review. I have add the write option document for Spark. And I find that we don't have Spark properties document. I want to raise another pr to add it after this pr is merged.

**Review by chenjunjiedada — APPROVED** [review]

**Comment by jerqi** [comment]
@pvary Could you help me review this pr?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java [hunk 2]
```diff
@@ -418,4 +418,64 @@ public String branch() {
...
+  }
+
+  public String parquetCompressionLevel() {
+    return confParser
+        .stringConf()
+        .option(SparkWriteOptions.COMPRESSION_LEVEL)
+        .sessionConf(SparkSQLProperties.COMPRESSION_LEVEL)
+        .tableProperty(TableProperties.PARQUET_COMPRESSION_LEVEL)
+        .defaultValue(TableProperties.PARQUET_COMPRESSION_LEVEL_DEFAULT)
+        .parseOptional();
```
Comment (ConeyLiu) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:439]: We can not get an optional value because we have set the default value, right?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java [hunk 3]
```diff
@@ -418,4 +418,64 @@ public String branch() {
...
+  }
+
+  public String avroCompressionLevel() {
+    return confParser
+        .stringConf()
+        .option(SparkWriteOptions.COMPRESSION_LEVEL)
+        .sessionConf(SparkSQLProperties.COMPRESSION_LEVEL)
+        .tableProperty(TableProperties.AVRO_COMPRESSION_LEVEL)
+        .defaultValue(TableProperties.AVRO_COMPRESSION_LEVEL_DEFAULT)
+        .parseOptional();
```
Comment (ConeyLiu) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:459]: Sames here

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestCompressionSettings {
+
+  private static final Configuration CONF = new Configuration();
```
Comment (ConeyLiu) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:75]: It would be better to cluster as follows:
```
private static final
private final
private
```

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        .format("iceberg")
+        .option(SparkWriteOptions.WRITE_FORMAT, format.toString())
+        .mode(SaveMode.Append)
+        .save(location.toString());
+
+    List<ManifestFile> manifestFiles = table.currentSnapshot().dataManifests(table.io());
+    try (ManifestReader<DataFile> reader = ManifestFiles.read(manifestFiles.get(0), table.io())) {
+      DataFile file = reader.iterator().next();
+      InputFile inputFile = table.io().newInputFile(file.path().toString());
+      Assert.assertEquals(
```
Comment (ConeyLiu) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:156]: Please follow the contribution guide using `AssertJ` instead. You can refer here: https://iceberg.apache.org/contribute/#assertj

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java (hunk 3 above)
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:459]: Good catch! Flink have a similar issue. I will raise another pr to correct the code of Flink.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java (hunk 2 above)
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:439]: Yes, good catch.

Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:439]: The default value is `null`, so we use the method `parseOptional`.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java (hunk 3 above)
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:459]: ditto.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java (hunk 4 above)
Comment (jerqi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:76]: Fixed.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java (hunk 5 above)
Comment (jerqi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java]: Fixed.

**Review by ConeyLiu — APPROVED** [review]
+1

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestCompressionSettings {
```
Comment (nastra) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:73]: The project is currently in the process of moving away from JUnit4, meaning that new tests should be written purely in Junit5 + using AssertJ assertions. See also https://iceberg.apache.org/contribute/#testing for some additional info

Comment (jerqi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:76]: Thanks for your review. I got it. I have migrated the test to the Junit 5 framework.

**Review by aokolnychyi — COMMENTED** [review]
Do we also have to cover other places that use `SparkFileWriterFactory`?
For instance, `SparkPositionDeltaWrite`?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java [hunk 7]
```diff
@@ -63,4 +63,9 @@ private SparkSQLProperties() {}
   // Controls the WAP branch used for write-audit-publish workflow.
   // When set, new snapshots will be committed to this branch.
   public static final String WAP_BRANCH = "spark.wap.branch";
+
+  // Controls write compress options
+  public static final String COMPRESSION_CODEC = "spark.sql.iceberg.write.compression-codec";
```
[the remark this answers is not in the document]
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkSQLProperties.java]: Fixed.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkFileWriterFactory.java [hunk 8]
```diff
@@ -76,6 +79,12 @@ class SparkFileWriterFactory extends BaseFileWriterFactory<InternalRow> {
     this.dataSparkType = dataSparkType;
     this.equalityDeleteSparkType = equalityDeleteSparkType;
     this.positionDeleteSparkType = positionDeleteSparkType;
+
```
[the remark this answers is not in the document]
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkFileWriterFactory.java]: Fixed.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java [hunk 9]
```diff
@@ -178,6 +185,32 @@ private WriterFactory createWriterFactory() {
     // broadcast the table metadata as the writer factory will be sent to executors
     Broadcast<Table> tableBroadcast =
         sparkContext.broadcast(SerializableTableWithSize.copyOf(table));
+    Map<String, String> writeProperties = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java]: Fixed.

**Comment by jerqi** [comment]
> Do we also have to cover other places that use `SparkFileWriterFactory`? For instance, `SparkPositionDeltaWrite`?

Thanks for you review, my mistake. I will add the modification of the code and add the ut.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    specMap.put(0, PartitionSpec.unpartitioned());
+    try (ManifestReader<DeleteFile> reader =
+        ManifestFiles.readDeleteManifest(deleteManifestFiles.get(0), table.io(), specMap)) {
+      DeleteFile file = reader.iterator().next();
+      InputFile inputFile = table.io().newInputFile(file.path().toString());
+      Assertions.assertThat(getCompressionType(inputFile))
+          .isEqualToIgnoringCase(properties.get(COMPRESSION_CODEC));
+    }
+
+    if (PARQUET.equals(format)) {
```
Comment (jerqi) [inline spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:185]: I find that we can't set table property for the action `RewritePositionDeletes`.  The action `RewritePositionDeletes` will always use parquet as the format.

Comment (jerqi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:185]: 企业微信截图_491440a5-40ce-404f-86c0-aec1addfe029
In my ut, I set the table property `DEFAULT_FILE_FORMAT` is orc, but the table properties is empty.

**Comment by jerqi** [comment]
> > Do we also have to cover other places that use `SparkFileWriterFactory`? For instance, `SparkPositionDeltaWrite`?
> 
> Thanks for you review, my mistake. I will add the modification of the code and add the ut.

@aokolnychyi I have addressed the comments.

**Comment by jerqi** [comment]
@nastra @aokolnychyi Gently ping. Sorry to bother you. Could you have another look if you have time?

**Comment by aokolnychyi** [comment]
Let me take another look.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.spark.SparkWriteConf;
+
+public class WritePropertiesUtil {
+
+  private WritePropertiesUtil() {}
+
+  public static Map<String, String> writeProperties(FileFormat format, SparkWriteConf writeConf) {
+    Map<String, String> writeProperties = Maps.newHashMap();
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java:38]: Optional: What about an empty line before the switch?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import static org.apache.iceberg.TableProperties.ORC_COMPRESSION_STRATEGY;
+import static org.apache.iceberg.TableProperties.PARQUET_COMPRESSION;
+import static org.apache.iceberg.TableProperties.PARQUET_COMPRESSION_LEVEL;
+
+import java.util.Map;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.spark.SparkWriteConf;
+
+public class WritePropertiesUtil {
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java:33]: I wonder whether it is worth adding a new util class in this case. Do we expect more related logic? I feel like we could just use `SparkWriteConf` and offer `writeProperties(FileFormat format)` method there, potentially making the newly added method like `parquetCompressionLevel` private. Any thoughts?

**Comment by aokolnychyi** [comment]
This change looks good to me. I am not sure about adding a new utility class vs just using `SparkWriteConf`.
@jerqi, what do you think?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java (hunk 12 above)
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java:33]: Good point. Thanks for your suggestion.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java (hunk 11 above)
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/WritePropertiesUtil.java:38]: OK, I got it.

**Comment by jerqi** [comment]
> This change looks good to me. I am not sure about adding a new utility class vs just using `SparkWriteConf`. @jerqi, what do you think?

The second solution seems more elegant. Thanks for your suggestion.  I have followed the suggestion.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java [hunk 13]
```diff
@@ -178,6 +178,7 @@ private WriterFactory createWriterFactory() {
     // broadcast the table metadata as the writer factory will be sent to executors
     Broadcast<Table> tableBroadcast =
         sparkContext.broadcast(SerializableTableWithSize.copyOf(table));
+
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java:181]: Minor: Is this change deliberate?

**Review by aokolnychyi — APPROVED** [review]
Looks good to me. Left a few suggestions.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java (hunk 10 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:185]: That seems like an issue worth looking into after this change.

cc @szehon-ho.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java [hunk 14]
```diff
@@ -104,6 +104,7 @@ class SparkPositionDeltaWrite implements DeltaWrite, RequiresDistributionAndOrde
   private final Context context;
 
   private boolean cleanupOnAbort = true;
+  private final Map<String, String> writeProperties;
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:107]: Minor: This variable should be co-located with other final variables above. It is super minor but let's do this in a follow-up.

**Comment by aokolnychyi** [comment]
Thanks, @jerqi! Thanks for reviewing, @ConeyLiu @chenjunjiedada.

**Comment by jerqi** [comment]
Thanks @aokolnychyi ! Code master. Thanks @ConeyLiu @chenjunjiedada @nastra , too.

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java (hunk 14 above)
Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java:107]: OK, I have raised a follow-up pr https://github.com/apache/iceberg/pull/8421.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java (hunk 10 above)
Comment (jerqi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:185]: OK, I will investigate it.

Comment (jerqi) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:185]: I find the root cause of this case.
The class `BaseMetadataTable` override the method `properties`
```
  @Override
  public Map<String, String> properties() {
    return ImmutableMap.of();
  }
```
Option 1: 
we can modify the class `BaseMetadataTable`.
https://github.com/apache/iceberg/pull/8428

Option2:
We can modify the class `PositionDeletesTable`
https://github.com/apache/iceberg/pull/8429

I prefer option 1. I feel the properties of  meta data table should respect the ones of  base table. We also don't have ways to modify the properites of  meta data table. @aokolnychyi @szehon-ho WDYT?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java [hunk 15]
```diff
@@ -418,4 +424,96 @@ public String branch() {
...
+
+  public Map<String, String> writeProperties(FileFormat format) {
+    Map<String, String> writeProperties = Maps.newHashMap();
+
+    switch (format) {
+      case PARQUET:
+        writeProperties.put(PARQUET_COMPRESSION, parquetCompressionCodec());
+        String parquetCompressionLevel = parquetCompressionLevel();
+        if (parquetCompressionLevel != null) {
+          writeProperties.put(PARQUET_COMPRESSION_LEVEL, parquetCompressionLevel);
```
Comment (szehon-ho) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:496]: Hi guys, while working on #8158 , I noticed the test failing and that this is missing delete file compression override.  I think the TestCompression delete file compression check passes by chance, I have a fix over there for it: https://github.com/apache/iceberg/pull/8299/commits/43e3cabb21964c7e2a9ebe424dcf4d1f03b8006c

Comment (jerqi) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:496]: Sorry for this mistake. I will check how to valid the correctness of the case.

File: spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java (hunk 10 above)
Comment (szehon-ho) [reply spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:185]: Hi, yea I am ok with option 1 as well,  @aokolnychyi what do you think?
